loader/src/LoaderEntry.cpp

```diff
@@ -121,6 +166,109 @@ void mira_escape(struct thread* td, void* uap)
 	printf("[-] mira_escape\n");
 }
 
+#define DIFFERENT_HASH 1
+#define SAME_HASH 0
+
+int MD5_hash_compare(const char *usbfile)
+{
+	 unsigned char c[16];
+	 unsigned char c2[16];
+    int i;
+    FILE *usb = fopen (usbfile, "rb");
+    FILE *hdd = fopen ("/user/MiraLoader.elf", "rb");
+    MD5_CTX mdContext;
+
+    MD5_CTX mdContext2;
+    int bytes2 = 0;
+    unsigned char data2[1024];
+
+    int bytes = 0;
+    unsigned char data[1024];
+
+    MD5_Init (&mdContext);
+    while ((bytes = fread (data, 1, 1024, usb)) != 0)
+        MD5_Update (&mdContext, data, bytes);
+    MD5_Final (c,&mdContext);
+   
+
+
+    MD5_Init (&mdContext2);
+    while ((bytes2 = fread (data2, 1, 1024, hdd)) != 0)
+        MD5_Update (&mdContext2, data2, bytes2);
+    MD5_Final (c2,&mdContext2);
+
+
+  printf("MD5 HASH OF USB ELF IS %i%i%i%i%i%i%i%i%i%i%i%i%i%i%i\n", c[1], c[2], c[3], c[4], c[5], c[6], c[7], c[8], c[9], c[10], c[11], c[12], c[13], c[14], c[15]);
+  printf("MD5 HASH OF HDD ELF IS %i%i%i%i%i%i%i%i%i%i%i%i%i%i%i\n", c2[1], c2[2], c2[3], c2[4], c2[5], c2[6], c2[7], c2[8], c2[9], c2[10], c2[11], c2[12], c2[13], c2[14], c2[15]);
+
+
+    for(i = 0; i < 16; i++) 
+    {
+    	printf("c[%i] = %i\n", i, c[i]);
+    	printf("c2[%i] = %i\n", i, c2[i]);
+         if(c[i] != c2[i])
+         {
+         	return DIFFERENT_HASH;
+         }
+
+
+    }
+
+    fclose (usb);
+    fclose (hdd);
+
+return SAME_HASH;
+}
+
+     int ftruncate(int fd, off_t length)
+     {
+     	return syscall(480, fd, length);
+     }
+
+
+    int munmap(void *addr,	size_t len)
+    {
+    	return syscall(73, addr, len);
+    }
+
+int copyFile(char *sourcefile)
+{
+
+    int sfd, dfd;
+    char *src, *dest;
+    size_t filesize;
+
+    /* SOURCE */
+    sfd = sceKernelOpen(sourcefile, O_RDONLY, 0x000);
+    filesize = sceKernelLseek(sfd, 0, SEEK_END);
+
+    if(filesize < 0) return -1;
+
+    src = _mmap(NULL, filesize, PROT_READ, MAP_PRIVATE, sfd, 0);
+
+    /* DESTINATION */
+    dfd = sceKernelOpen("/user/MiraLoader.elf", O_RDWR | O_CREAT | O_TRUNC, 0666);
+
+    ftruncate(dfd, filesize);
```

Comment:
Is this needed if you are passing O_TRUNC?

Reply:
O_TRUNC 0s it and ftruncate makes the file the correct size iirc